File: v1/sdks/go-sdk/client_test.go
package nexus

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
)

func TestClientRunToolSendsHeadersAndParsesResponse(t *testing.T) {
	t.Parallel()

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/run" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		if got := r.Header.Get("X-NEXUS-CORE-KEY"); got != "test-key" {
			t.Fatalf("expected API key header, got %q", got)
		}
		if got := r.Header.Get("Idempotency-Key"); got != "idem-1" {
			t.Fatalf("expected idempotency header, got %q", got)
		}
		if got := r.Header.Get("X-Timeout-Ms"); got != "750" {
			t.Fatalf("expected timeout header, got %q", got)
		}

		var body RunRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatalf("decode request: %v", err)
		}
		if body.ToolName != "echo" {
			t.Fatalf("expected tool echo, got %s", body.ToolName)
		}

		_ = json.NewEncoder(w).Encode(RunResponse{
			RequestID: "req-1",
			Decision:  "allow",
			ToolName:  "echo",
			Status:    "success",
			Result: map[string]any{
				"ok": true,
			},
			LatencyMS: 12,
		})
	}))
	defer srv.Close()

	client := NewClient(srv.URL, "test-key")
	resp, err := client.RunTool(context.Background(), RunRequest{
		ToolName: "echo",
		Input:    map[string]any{"message": "hello"},
		Headers: map[string]string{
			"Idempotency-Key": "idem-1",
			"X-Timeout-Ms":    "750",
		},
	})
	if err != nil {
		t.Fatalf("RunTool error: %v", err)
	}
	if resp.Status != "success" || resp.RequestID != "req-1" {
		t.Fatalf("unexpected response: %+v", resp)
	}
}

func TestClientListToolsReturnsItems(t *testing.T) {
	t.Parallel()

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/tools" {
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
		_ = json.NewEncoder(w).Encode(map[string]any{
			"items": []Tool{
				{
					ID:         "tool-1",
					Name:       "echo",
					Kind:       "http",
					Method:     "POST",
					URL:        "http://example.test/echo",
					ActionType: "read",
					Enabled:    true,
				},
			},
		})
	}))
	defer srv.Close()

	client := NewClient(srv.URL, "test-key")
	items, err := client.ListTools(context.Background())
	if err != nil {
		t.Fatalf("ListTools error: %v", err)
	}
	if len(items) != 1 || items[0].Name != "echo" {
		t.Fatalf("unexpected items: %+v", items)
	}
}

func TestClientReturnsAPIErrorEnvelope(t *testing.T) {
	t.Parallel()

	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(http.StatusForbidden)
		_ = json.NewEncoder(w).Encode(map[string]any{
			"error": map[string]string{
				"code":    "POLICY_DENIED",
				"message": "blocked by policy",
			},
		})
	}))
	defer srv.Close()

	client := NewClient(srv.URL, "test-key")
	_, err := client.ListTools(context.Background())
	if err == nil {
		t.Fatal("expected error")
	}
	if !strings.Contains(err.Error(), "POLICY_DENIED: blocked by policy") {
		t.Fatalf("unexpected error: %v", err)
	}
}

File: v1/sdks/typescript-sdk/src/client.test.ts
import { describe, expect, it, vi } from 'vitest';

import { NexusClient } from './client.js';
import { NexusError } from './types.js';

function jsonResponse(body: unknown, init: ResponseInit = {}): Response {
  return new Response(JSON.stringify(body), {
    status: init.status ?? 200,
    headers: {
      'content-type': 'application/json',
      ...(init.headers ?? {}),
    },
  });
}

describe('NexusClient', () => {
  it('sends auth headers and decodes tool lists', async () => {
    const fetchMock = vi.fn<typeof fetch>().mockResolvedValue(
      jsonResponse({
        items: [
          {
            id: 'tool-1',
            name: 'echo',
            kind: 'http',
            method: 'POST',
            url: 'http://example.test/echo',
            action_type: 'read',
            enabled: true,
          },
        ],
      }),
    );

    const client = new NexusClient({
      baseUrl: 'http://localhost:8080/',
      apiKey: 'key-1',
      actor: 'agent-1',
      role: 'bot',
      scopes: 'tools:run',
      fetch: fetchMock,
    });

    const tools = await client.listTools();

    expect(tools).toHaveLength(1);
    expect(tools[0]?.name).toBe('echo');
    expect(fetchMock).toHaveBeenCalledWith(
      'http://localhost:8080/v1/tools',
      expect.objectContaining({
        method: 'GET',
        headers: expect.objectContaining({
          'Content-Type': 'application/json',
          'X-NEXUS-CORE-KEY': 'key-1',
          'X-NEXUS-ACTOR': 'agent-1',
          'X-NEXUS-ROLE': 'bot',
          'X-NEXUS-SCOPES': 'tools:run',
        }),
      }),
    );
  });

  it('maps 429 responses to blocked run payloads', async () => {
    const fetchMock = vi.fn<typeof fetch>().mockResolvedValue(
      jsonResponse(
        {
          error: {
            code: 'RATE_LIMITED',
            message: 'tenant run rate limit exceeded',
          },
        },
        { status: 429 },
      ),
    );

    const client = new NexusClient({
      baseUrl: 'http://localhost:8080',
      fetch: fetchMock,
    });

    const result = await client.run('echo', { message: 'hello' });

    expect(result).toEqual({
      request_id: '',
      decision: 'deny',
      tool_name: 'echo',
      status: 'blocked',
      error: {
        code: 'RATE_LIMITED',
        message: 'tenant run rate limit exceeded',
      },
      latency_ms: 0,
    });
  });

  it('throws NexusError for non-blocking server failures', async () => {
    const fetchMock = vi.fn<typeof fetch>().mockResolvedValue(
      jsonResponse(
        {
          error: {
            code: 'INTERNAL',
            message: 'boom',
          },
        },
        { status: 500 },
      ),
    );

    const client = new NexusClient({
      baseUrl: 'http://localhost:8080',
      fetch: fetchMock,
    });

    await expect(client.listTools()).rejects.toEqual(new NexusError(500, 'INTERNAL', 'boom'));
  });
});
